matminer/featurizers/function.py

```diff
@@ -0,0 +1,229 @@
+
+        if inplace:
+            for k in new_col_names:
+                df[k] = new_df[k]
+            return df
+        else:
+            return new_df
+
+    def featurize(self, *args):
+        """
+        Main featurizer function, essentially iterates over all
+        of the functions in self.function_list to generate
+        features for each argument.
+
+        Args:
+            *args: list of numbers to generate functional output
+                features
+
+        Returns:
+            list of functional outputs corresponding to input args
+        """
+        return list(self._exp_iter(*args, postprocess=self.postprocess))
+
+    def feature_labels(self):
+        """
+        Returns:
+            Set of feature labels corresponding to expressions
+        """
+        return None
+
+    def generate_string_expressions(self, input_variable_names):
+        """
+        Method to generate string expressions for input strings,
+        mainly used to generate columns names for featurize_dataframe
+
+        Args:
+            input_variable_names ([str]): strings corresponding to
+                functional input variable names
+
+        Returns:
+            List of string expressions generated by substitution of
+            variable names into functions
+        """
+        if isinstance(input_variable_names, string_types):
+            input_variable_names = [input_variable_names]
+        postprocess = sp.latex if self.latexify_labels else str
+        return list(self._exp_iter(*input_variable_names,
+                                   postprocess=postprocess))
+
+    def _exp_iter(self, *args, postprocess=None):
+        """
+        Generates an iterator for substitution of a set
+        of args into the set of expression corresponding
+        to the featurizer, intended primarily to remove
+        replicated code in featurize and feature labels
+
+        Args:
+            *args: args to loop over combinations and substitions for
+            postprocess (function): postprocessing function, e. g.
+                to cast to another type, float, str
+
+        Returns:
+            iterator for all substituted expressions
+
+        """
+        postprocess = postprocess or (lambda x: x)
+        for n in range(1, self.multi_feature_depth + 1):
+            for arg_combo in itertools.combinations(args, n):
+                subs_dict = {"x{}".format(m): arg
+                             for m, arg in enumerate(arg_combo)}
+                for exp in self.exp_dict[n]:
+                    # TODO: this is a workaround for the problem
+                    # TODO: postprocessing functional incompatility,
+                    # TODO: e. g. sqrt(-1), 1 / 0
+                    try:
+                        yield postprocess(exp.subs(subs_dict))
+                    except (TypeError, ValueError):
+                        yield None
+
+    def citations(self):
+        return ["@article{Ramprasad2017,"
```

Comment:
I think we should cite Ghiringhelli et al. in place of or in addition to this reference.